File: day02_p2.py
def parse_game_data(line):
    game_id, data = line.split(': ')
    game_id = int(game_id.split(' ')[1])
    reveals = data.split('; ')

    game_reveals = []
    for reveal in reveals:
        counts = {'red': 0, 'green': 0, 'blue': 0}
        cubes = reveal.split(', ')
        for cube in cubes:
            parts = cube.split(' ')
            if len(parts) == 2:
                count, color = parts
                counts[color] = max(counts[color], int(count))
        game_reveals.append(counts)

    return game_id, game_reveals


def calculate_min_cubes(game_reveals):
    min_cubes = {'red': 0, 'green': 0, 'blue': 0}
    for reveal in game_reveals:
        for color in min_cubes:
            min_cubes[color] = max(min_cubes[color], reveal.get(color, 0))
    return min_cubes


def calculate_power(min_cubes):
    return min_cubes['red'] * min_cubes['green'] * min_cubes['blue']


def main(input_file):
    with open(input_file, 'r') as file:
        games = [parse_game_data(line.strip()) for line in file]

    total_power = 0
    for game_id, reveals in games:
        min_cubes = calculate_min_cubes(reveals)
        total_power += calculate_power(min_cubes)

    return total_power


input_file = 'input.txt'
print(main(input_file))

File: day02_p2.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::collections::HashMap;

fn parse_game_data(line: &str) -> (i32, Vec<HashMap<String, i32>>) {
    let parts: Vec<&str> = line.split(": ").collect();
    let game_id = parts[0].split(" ").collect::<Vec<&str>>()[1].parse::<i32>().unwrap();
    let reveals = parts[1].split("; ")
        .map(|reveal| {
            let mut counts = HashMap::new();
            counts.insert("red".to_string(), 0);
            counts.insert("green".to_string(), 0);
            counts.insert("blue".to_string(), 0);
            for cube in reveal.split(", ") {
                let cube_parts: Vec<&str> = cube.split(" ").collect();
                if cube_parts.len() == 2 {
                    let count = cube_parts[0].parse::<i32>().unwrap();
                    let color = cube_parts[1].to_string();
                    *counts.entry(color).or_insert(0) = std::cmp::max(counts[&color], count);
                }
            }
            counts
        }).collect();
    
    (game_id, reveals)
}

fn calculate_min_cubes(reveals: &Vec<HashMap<String, i32>>) -> HashMap<String, i32> {
    let mut min_cubes = HashMap::new();
    min_cubes.insert("red".to_string(), 0);
    min_cubes.insert("green".to_string(), 0);
    min_cubes.insert("blue".to_string(), 0);

    for reveal in reveals {
        for (color, &count) in reveal {
            let entry = min_cubes.entry(color.to_string()).or_insert(0);
            *entry = std::cmp::max(*entry, count);
        }
    }
    min_cubes
}

fn calculate_power(min_cubes: &HashMap<String, i32>) -> i32 {
    min_cubes.get("red").unwrap_or(&0) * min_cubes.get("green").unwrap_or(&0) * min_cubes.get("blue").unwrap_or(&0)
}

fn main() -> io::Result<()> {
    let path = Path::new("input.txt"); // Adjust the path as needed
    let file = File::open(&path)?;
    let reader = io::BufReader::new(file);

    let mut total_power = 0;
    for line in reader.lines() {
        let line = line?;
        let (_game_id, reveals) = parse_game_data(&line);
        let min_cubes = calculate_min_cubes(&reveals);
        total_power += calculate_power(&min_cubes);
    }

    println!("{}", total_power);
    Ok(())
}
